source/simulator/core.cc

```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                  test_newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  test_newton_residual = std::sqrt(test_newton_residual_velo * test_newton_residual_velo
+                                                   + test_newton_residual_pres * test_newton_residual_pres);
+
+                  if (test_newton_residual < (1.0 - alpha * lambda) * newton_residual
+                      ||
+                      line_search_iteration >= parameters.max_newton_line_search_iterations
+                      ||
+                      use_picard)
+                    {
+                      pcout << "      Total relative residual after nonlinear iteration " << nonlinear_iteration+1
+                            << ": " << newton_residual/initial_newton_residual << ", norm of the rhs: " << test_newton_residual
+                            << ", newton_derivative_scaling_factor: " << parameters.newton_derivative_scaling_factor << std::endl;
+                      break;
+                    }
+                  else
+                    {
+
+                      pcout << "   Line search iteration " << line_search_iteration << ", with norm of the rhs "
+                            << test_newton_residual << " and going to " << (1.0 - alpha * lambda) * newton_residual
+                            << ", relative residual: " << newton_residual/initial_newton_residual << std::endl;
+
+                      /**
+                       * The line search step was not sufficient to decrease the residual
+                       * enough, so we take a smaller step to see if it improves the residual.
+                       */
+                      lambda *= (2.0/3.0);// TODO: make a parameter out of this.
+                    }
+
+                  line_search_iteration++;
+                  Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
+                         ExcMessage ("This tests the while condition. This condition should "
+                                     "actually never be false, because the break statement "
+                                     "above should have caught it."));
+                }
+              while (line_search_iteration <= parameters.max_newton_line_search_iterations);
+              // The while condition should actually never be false, because the break statement above should have caught it.
+
+
+              if (use_picard == true)
+                {
+                  parameters.switch_initial_newton_residual = newton_residual;
+                  parameters.newton_residual = newton_residual;
+                }
+              else
+                {
+                  /**
+                   * This method allows to slowly introduce the derivatives based
+                   * on the improvement of the residual.
+                   */
+                  if (parameters.use_newton_residual_scaling_method) //said's method
```

Comment:
Does this comment mean: *Said*'s method? Please move it into the longer comment above and mention where to find this method.